navigation/perfect_memory_base.h

```diff
@@ -57,65 +57,65 @@ class PerfectMemoryBase
     //------------------------------------------------------------------------
     // Public API
     //------------------------------------------------------------------------
-    virtual void train(const cv::Mat &image, bool saveImage = false) override
+    virtual void train(const cv::Mat &image) override
     {
         const auto &unwrapRes = getUnwrapResolution();
         assert(image.cols == unwrapRes.width);
         assert(image.rows == unwrapRes.height);
         assert(image.type() == CV_8UC1);
 
         // Add snapshot
-        const size_t index = addSnapshot(image);
-
-        // Write image to disk, if desired
-        if (saveImage) {
-            saveSnapshot(index, image);
-        }
+        addSnapshot(image);
     }
 
-    //! Get differences between image and stored snapshots
-    std::vector<std::vector<float>> getImageDifferences(const cv::Mat &image) const
+    /*!
+     * \brief Get differences between current view and stored snapshots
+     *
+     * The parameters are perfect-forwarded to the Rotater class, so e.g. for
+     * InSilicoRotater one passes in a cv::Mat and (optionally) an unsigned int
+     * for the scan step and for the AntWorldRotater, one passes in one or more
+     * angles.
+     */
+    template<class... Ts>
+    std::vector<std::vector<float>> getImageDifferences(Ts &&... args) const
     {
-        const auto &unwrapRes = getUnwrapResolution();
-        assert(image.cols == unwrapRes.width);
-        assert(image.rows == unwrapRes.height);
-        assert(image.type() == CV_8UC1);
         const size_t numSnapshots = getNumSnapshots();
         assert(numSnapshots > 0);
-        const size_t scanStep = getScanStep();
 
-        // Clone mask and image so they can be rolled in place
-        getMaskImage().copyTo(m_ScratchMaskImage);
-        image.copyTo(m_ScratchRollImage);
+        Rotater rotater(getUnwrapResolution(), getMaskImage(), std::forward<Ts>(args)...);
 
         // Create vector to store RIDF values
         std::vector<std::vector<float>> differences(numSnapshots);
         for (auto &d : differences) {
-            d.resize(m_ScratchRollImage.cols);
+            d.resize(rotater.max());
         }
 
         // Scan across image columns
-        for (int col = 0; col < m_ScratchRollImage.cols; col += scanStep) {
+        rotater([this, &differences, numSnapshots](const cv::Mat &fr, const cv::Mat &mask, size_t i) {
```

Comment:
Personally I prefer indenting this like:
```c++
rotater(
    [this, &differences, numSnapshots](const cv::Mat &fr, const cv::Mat &mask, size_t i)
    {
        ....
    });
```
```
So the lambda function is intended like a normal function but that is a **very** minor nit-pick

Reply:
Also I don't think there's any need to use ``operator ()`` for this operation - I'd just give it a ``rotate`` method

Reply:
Makes sense